# syntax-transform/3.3/0001-space-before-argument-parentheses.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, when calling a function with more than one argument, it is
# allowed to have a space between the function name and the `(` but shows the
# warning "don't put space before argument parentheses". In Ruby versions
# newer than 1.8, this syntax causes the expression after the function name to
# be interpreted as a single argument instead. As a result, this code snippet
# works in Ruby 1.8 but results in a syntax error in Ruby versions newer than
# 1.8 because the `(1, 2)` gets interpreted as a single argument.
#
#   def f(x, y)
#     p x
#     p y
#   end
#   f (1, 2)
#
# However, Ruby 1.8 uses the modern parsing behavior for certain types of
# expressions, non-recursively, with the exception that no syntax error occurs
# when a function is called with more than one argument with a space before the
# parentheses:
#
# * The expressions on the inside of a pair of parentheses
# * The expressions inside of a block
# * The expressions inside of a class definition
# * The expressions inside of a function definition#
#
# This snippet prints `12`, `22`, `12` and `22` in Ruby 1.8 but `22`, `22`, `22`
# and `22` in Ruby versions newer than 1.8.
#
#   def f(x)
#     x * 2
#   end
#   p f (1) + 10
#   p(f (1) + 10)
#   a = f (1) + 10
#   p a
#   a = (f (1) + 10)
#   p a
#
# This code snippet is a test of a bunch of edge cases in the parser to check
# for consistency with the Ruby 1.8 behavior.
#
#   def f(x)
#     not x
#   end
#   a = f (true) || true
#   if a
#     p 'test 1 ok'
#   else
#     p 'test 1 failed'
#   end
#   a = (f (true) || true)
#   unless a
#     p 'test 2 ok'
#   else
#     p 'test 2 failed'
#   end
#   unless (f (true) || true)
#     p 'test 3 ok'
#   else
#     p 'test 3 failed'
#   end
#   unless lambda {f (true) || true}.call
#     p 'test 4 ok'
#   else
#     p 'test 4 failed'
#   end
#   if lambda {return f (true) || true}.call
#     p 'test 5 ok'
#   else
#     p 'test 5 failed'
#   end
#   if lambda {true and f (true) || true}.call
#     p 'test 6 ok'
#   else
#     p 'test 6 failed'
#   end
#   unless lambda {f (true) || true and true}.call
#     p 'test 7 ok'
#   else
#     p 'test 7 failed'
#   end
#   a = class << nil
#     f (true) || true
#   end
#   unless a
#     p 'test 8 ok'
#   else
#     p 'test 8 failed'
#   end
#   if f (true) || true # The behavior here is caused by Ruby bug #6347, fixed in Ruby 2.0
#     p 'test 9 ok'
#   else
#     p 'test 9 failed'
#   end
#   if "#{f (true) || true}" == 'true' # The behavior here is caused by Ruby bug #6347, fixed in Ruby 2.0
#     p 'test 10 ok'
#   else
#     p 'test 10 failed'
#   end
#   class << f (true) || true # The behavior here is caused by Ruby bug #6347, fixed in Ruby 2.0
#     if self == TrueClass
#       p 'test 11 ok'
#     else
#       p 'test 11 failed'
#     end
#   end
#   begin
#     class A < f (true) || true; end # The behavior here is caused by Ruby bug #6347, fixed in Ruby 2.0
#   rescue TypeError => e
#     if e.message.include? 'TrueClass'
#       p 'test 12 ok'
#     else
#       p 'test 12 failed'
#     end
#   end
#   unless lambda {not (true) || true}.call # Be careful: the expression type of the `not` keyword changed in Ruby 1.9
#     p 'test 13 ok'
#   else
#     p 'test 13 failed'
#   end

--- a/parse.y
+++ b/parse.y
@@ -798,6 +798,8 @@ parser_token2id(struct parser_params *p, enum yytokentype tok)
       TOKEN2ID(k__END__);
       TOKEN2ID(tLOWEST);
       TOKEN2ID(tUMINUS_NUM);
+      TOKEN2ID(tMKXP_LEGACY_LPAREN_ARG);
+      TOKEN2ID(tMKXP_LEGACY_STAR);
       TOKEN2ID(tLAST_TOKEN);
 #undef TOKEN2ID
 #undef TOKEN2ID2
@@ -2107,6 +2109,7 @@ get_nd_args(struct parser_params *p, NODE *node)
 %type <node> p_args p_args_head p_args_tail p_args_post p_arg p_rest
 %type <node> p_value p_primitive p_variable p_var_ref p_expr_ref p_const
 %type <node> p_kwargs p_kwarg p_kw
+%type <node> mkxp_legacy_expr mkxp_legacy_call_args2 mkxp_open_args
 %type <id>   keyword_variable user_variable sym operation operation2 operation3
 %type <id>   cname fname op f_rest_arg f_block_arg opt_f_block_arg f_norm_arg f_bad_arg
 %type <id>   f_kwrest f_label f_arg_asgn call_op call_op2 reswords relop dot_or_colon
@@ -2178,6 +2181,9 @@ get_nd_args(struct parser_params *p, NODE *node)
 %token tIGNORED_NL tCOMMENT tEMBDOC_BEG tEMBDOC tEMBDOC_END
 %token tHEREDOC_BEG tHEREDOC_END k__END__
 
+%token tMKXP_LEGACY_LPAREN_ARG	"( arg legacy"
+%token tMKXP_LEGACY_STAR	"* legacy"
+
 /*
  *	precedence table
  */
@@ -2237,6 +2243,121 @@ program		:  {
                     }
                 ;
 
+mkxp_star       : tSTAR
+                | tMKXP_LEGACY_STAR
+                ;
+
+mkxp_legacy_expr: command_call
+                | mkxp_legacy_expr keyword_and mkxp_legacy_expr
+                    {
+                        $$ = logop(p, idAND, $1, $3, &@2, &@$);
+                    }
+                | mkxp_legacy_expr keyword_or mkxp_legacy_expr
+                    {
+                        $$ = logop(p, idOR, $1, $3, &@2, &@$);
+                    }
+                | keyword_not mkxp_legacy_expr
+                    {
+                        $$ = call_uni_op(p, method_cond(p, $2, &@2), METHOD_NOT, &@1, &@$);
+                    }
+                | '!' command_call
+                    {
+                        $$ = call_uni_op(p, method_cond(p, $2, &@2), '!', &@1, &@$);
+                    }
+                | arg
+                ;
+
+mkxp_legacy_call_args2
+                : arg_value ',' args opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = arg_blk_pass(rest_arg_append(p, NEW_LIST($1, &@1), $3, &@$), $4);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add_star!(args_add!(args_new!, $1), $3), $4) %*/
+                    }
+                | arg_value ',' block_arg
+                    {
+                    /*%%%*/
+                        $$ = arg_blk_pass(NEW_LIST($1, &@1), $3);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_new!, $1), $3) %*/
+                    }
+                | assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = $1 ? NEW_LIST(new_hash(p, $1, &@1), &@1) : 0;
+                        $$ = arg_blk_pass($$, $2);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_new!, bare_assoc_hash!($1)), $2) %*/
+                    }
+                | assocs ',' tMKXP_LEGACY_STAR arg_value opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = $1 ? rest_arg_append(p, NEW_LIST(new_hash(p, $1, &@1), &@1), $4, &@$) : NEW_SPLAT($4, &@$);
+                        $$ = arg_blk_pass($$, $5);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add_star!(args_add!(args_new!, bare_assoc_hash!($1)), $4), $5) %*/
+                    }
+                | arg_value ',' assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = NEW_LIST($1, &@1);
+                        $$ = $3 ? last_arg_append(p, $$, new_hash(p, $3, &@3), &@$) : $$;
+                        $$ = arg_blk_pass($$, $4);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_add!(args_new!, $1), bare_assoc_hash!($3)), $4) %*/
+                    }
+                | arg_value ',' args ',' assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = rest_arg_append(p, NEW_LIST($1, &@1), $3, &@$);
+                        $$ = $5 ? arg_append(p, $$, new_hash(p, $5, &@5), &@$) : $$;
+                        $$ = arg_blk_pass($$, $6);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_add_star!(args_add!(args_new!, $1), $3), bare_assoc_hash!($5)), $6) %*/
+                    }
+                | arg_value ',' assocs ',' tMKXP_LEGACY_STAR arg_value opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = NEW_LIST($1, &@1);
+                        $$ = $3 ? last_arg_append(p, $$, new_hash(p, $3, &@3), &@$) : $$;
+                        $$ = arg_blk_pass(rest_arg_append(p, $$, $6, &@$), $7);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add_star!(args_add!(args_add!(args_new!, $1), bare_assoc_hash!($3)), $6), $7) %*/
+                    }
+                | arg_value ',' args ',' assocs ',' tMKXP_LEGACY_STAR arg_value opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = rest_arg_append(p, NEW_LIST($1, &@1), $3, &@$);
+                        $$ = $5 ? arg_append(p, $$, new_hash(p, $5, &@5), &@$) : $$;
+                        $$ = arg_blk_pass(rest_arg_append(p, $$, $8, &@$), $9);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add_star!(args_add!(args_add_star!(args_add!(args_new!, $1), $3), bare_assoc_hash!($5)), $8), $9) %*/
+                    }
+                | tMKXP_LEGACY_STAR arg_value opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = arg_blk_pass(NEW_SPLAT($2, &@$), $3);
+                    /*% %*/
+                    /*% ripper[brace]: args_add_block!(args_add_star!(args_new!, $2), $3) %*/
+                    }
+                | block_arg
+                    /*% ripper[brace]: args_add_block!(args_new!, $1) %*/
+                ;
+
+mkxp_open_args  : call_args
+                | tMKXP_LEGACY_LPAREN_ARG {SET_LEX_STATE(EXPR_ENDARG);} ')'
+                    {
+                        rb_warn0("don't put space before argument parentheses");
+                        $$ = 0;
+                    }
+                | tMKXP_LEGACY_LPAREN_ARG mkxp_legacy_call_args2 {SET_LEX_STATE(EXPR_ENDARG);} ')'
+                    {
+                        rb_warn0("don't put space before argument parentheses");
+                        $$ = $2;
+                    }
+                ;
+
 top_compstmt	: top_stmts opt_terms
                     {
                         $$ = void_stmts(p, $1);
@@ -2928,56 +3049,56 @@ mlhs_basic	: mlhs_head
                     /*% %*/
                     /*% ripper: mlhs_add!($1, $2) %*/
                     }
-                | mlhs_head tSTAR mlhs_node
+                | mlhs_head mkxp_star mlhs_node
                     {
                     /*%%%*/
                         $$ = NEW_MASGN($1, $3, &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_star!($1, $3) %*/
                     }
-                | mlhs_head tSTAR mlhs_node ',' mlhs_post
+                | mlhs_head mkxp_star mlhs_node ',' mlhs_post
                     {
                     /*%%%*/
                         $$ = NEW_MASGN($1, NEW_POSTARG($3,$5,&@$), &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_post!(mlhs_add_star!($1, $3), $5) %*/
                     }
-                | mlhs_head tSTAR
+                | mlhs_head mkxp_star
                     {
                     /*%%%*/
                         $$ = NEW_MASGN($1, NODE_SPECIAL_NO_NAME_REST, &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_star!($1, Qnil) %*/
                     }
-                | mlhs_head tSTAR ',' mlhs_post
+                | mlhs_head mkxp_star ',' mlhs_post
                     {
                     /*%%%*/
                         $$ = NEW_MASGN($1, NEW_POSTARG(NODE_SPECIAL_NO_NAME_REST, $4, &@$), &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_post!(mlhs_add_star!($1, Qnil), $4) %*/
                     }
-                | tSTAR mlhs_node
+                | mkxp_star mlhs_node
                     {
                     /*%%%*/
                         $$ = NEW_MASGN(0, $2, &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_star!(mlhs_new!, $2) %*/
                     }
-                | tSTAR mlhs_node ',' mlhs_post
+                | mkxp_star mlhs_node ',' mlhs_post
                     {
                     /*%%%*/
                         $$ = NEW_MASGN(0, NEW_POSTARG($2,$4,&@$), &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_post!(mlhs_add_star!(mlhs_new!, $2), $4) %*/
                     }
-                | tSTAR
+                | mkxp_star
                     {
                     /*%%%*/
                         $$ = NEW_MASGN(0, NODE_SPECIAL_NO_NAME_REST, &@$);
                     /*% %*/
                     /*% ripper: mlhs_add_star!(mlhs_new!, Qnil) %*/
                     }
-                | tSTAR ',' mlhs_post
+                | mkxp_star ',' mlhs_post
                     {
                     /*%%%*/
                         $$ = NEW_MASGN(0, NEW_POSTARG(NODE_SPECIAL_NO_NAME_REST, $3, &@$), &@$);
@@ -3253,6 +3374,7 @@ op		: '|'		{ ifndef_ripper($$ = '|'); }
                 | '-'		{ ifndef_ripper($$ = '-'); }
                 | '*'		{ ifndef_ripper($$ = '*'); }
                 | tSTAR		{ ifndef_ripper($$ = '*'); }
+                | tMKXP_LEGACY_STAR		{ ifndef_ripper($$ = '*'); }
                 | '/'		{ ifndef_ripper($$ = '/'); }
                 | '%'		{ ifndef_ripper($$ = '%'); }
                 | tPOW		{ ifndef_ripper($$ = tPOW); }
@@ -3739,14 +3861,14 @@ command_args	:   {
                          */
                         int lookahead = 0;
                         switch (yychar) {
-                          case '(': case tLPAREN: case tLPAREN_ARG: case '[': case tLBRACK:
+                          case '(': case tLPAREN: case tLPAREN_ARG: case tMKXP_LEGACY_LPAREN_ARG: case '[': case tLBRACK:
                             lookahead = 1;
                         }
                         if (lookahead) CMDARG_POP();
                         CMDARG_PUSH(1);
                         if (lookahead) CMDARG_PUSH(0);
                     }
-                  call_args
+                  mkxp_open_args
                     {
                         /* call_args can be followed by tLBRACE_ARG (that does CMDARG_PUSH(0) in the lexer)
                          * but the push must be done after CMDARG_POP() in the parser.
@@ -3825,11 +3947,11 @@ args		: arg_value
                 ;
 
 /* value */
-arg_splat	: tSTAR arg_value
+arg_splat	: mkxp_star arg_value
                     {
                         $$ = $2;
                     }
-                | tSTAR /* none */
+                | mkxp_star /* none */
                     {
                         forwarding_arg_check(p, idFWD_REST, idFWD_ALL, "rest");
                     /*%%%*/
@@ -3852,14 +3974,14 @@ mrhs		: args ',' arg_value
                     /*% %*/
                     /*% ripper: mrhs_add!(mrhs_new_from_args!($1), $3) %*/
                     }
-                | args ',' tSTAR arg_value
+                | args ',' mkxp_star arg_value
                     {
                     /*%%%*/
                         $$ = rest_arg_append(p, $1, $4, &@$);
                     /*% %*/
                     /*% ripper: mrhs_add_star!(mrhs_new_from_args!($1), $4) %*/
                     }
-                | tSTAR arg_value
+                | mkxp_star arg_value
                     {
                     /*%%%*/
                         $$ = NEW_SPLAT($2, &@$);
@@ -3908,6 +4030,14 @@ primary		: literal
                     /*% %*/
                     /*% ripper: paren!($2) %*/
                     }
+                | tMKXP_LEGACY_LPAREN_ARG mkxp_legacy_expr {SET_LEX_STATE(EXPR_ENDARG);} ')'
+                    {
+                    /*%%%*/
+                        if (nd_type_p($2, NODE_SELF)) RNODE_SELF($2)->nd_state = 0;
+                        $$ = $2;
+                    /*% %*/
+                    /*% ripper: paren!($2) %*/
+                    }
                 | tLPAREN compstmt ')'
                     {
                     /*%%%*/
@@ -4565,7 +4695,7 @@ f_margs		: f_marg_list
                     }
                 ;
 
-f_rest_marg	: tSTAR f_norm_arg
+f_rest_marg	: mkxp_star f_norm_arg
                     {
                     /*%%%*/
                         $$ = assignable(p, $2, 0, &@$);
@@ -4573,7 +4703,7 @@ f_rest_marg	: tSTAR f_norm_arg
                     /*% %*/
                     /*% ripper: assignable(p, $2) %*/
                     }
-                | tSTAR
+                | mkxp_star
                     {
                     /*%%%*/
                         $$ = NODE_SPECIAL_NO_NAME_REST;
@@ -5013,7 +5143,7 @@ case_args	: arg_value
                     /*% %*/
                     /*% ripper: args_add!(args_new!, $1) %*/
                     }
-                | tSTAR arg_value
+                | mkxp_star arg_value
                     {
                     /*%%%*/
                         $$ = NEW_SPLAT($2, &@$);
@@ -5028,7 +5158,7 @@ case_args	: arg_value
                     /*% %*/
                     /*% ripper: args_add!($1, $3) %*/
                     }
-                | case_args ',' tSTAR arg_value
+                | case_args ',' mkxp_star arg_value
                     {
                     /*%%%*/
                         $$ = rest_arg_append(p, $1, $4, &@$);
@@ -5325,7 +5455,7 @@ p_find		: p_rest ',' p_args_post ',' p_rest
                 ;
 
 
-p_rest		: tSTAR tIDENTIFIER
+p_rest		: mkxp_star tIDENTIFIER
                     {
                     /*%%%*/
                         error_duplicate_pattern_variable(p, $2, &@2);
@@ -5333,7 +5463,7 @@ p_rest		: tSTAR tIDENTIFIER
                     /*% %*/
                     /*% ripper: assignable(p, var_field(p, $2)) %*/
                     }
-                | tSTAR
+                | mkxp_star
                     {
                     /*%%%*/
                         $$ = 0;
@@ -6099,7 +6229,9 @@ backref		: tNTH_REF
 superclass	: '<'
                     {
                         SET_LEX_STATE(EXPR_BEG);
-                        p->command_start = TRUE;
+                        if (!mkxp_parser_is_syntax_transform_active(p, 1, 9, -1)) {
+                            p->command_start = TRUE;
+                        }
                     }
                   expr_value term
                     {
@@ -6530,7 +6662,7 @@ f_optarg	: f_opt
                 ;
 
 restarg_mark	: '*'
-                | tSTAR
+                | mkxp_star
                 ;
 
 f_rest_arg	: restarg_mark tIDENTIFIER
@@ -8551,7 +8683,9 @@ parser_peek_variable_name(struct parser_params *p)
         break;
       case '{':
         p->lex.pcur = ptr;
-        p->command_start = TRUE;
+        if (!mkxp_parser_is_syntax_transform_active(p, 1, 9, -1)) {
+            p->command_start = TRUE;
+        }
         return tSTRING_DBEG;
       default:
         return 0;
@@ -10377,8 +10511,8 @@ parse_ident(struct parser_params *p, int c, int cmd_state)
                 set_yylval_name(rb_intern2(tok(p), toklen(p)));
                 return kw->id[0];
             }
-            SET_LEX_STATE(kw->state);
-            if (IS_lex_state(EXPR_BEG)) {
+            SET_LEX_STATE(mkxp_parser_is_syntax_transform_active(p, 1, 8, -1) && kw->id[0] == keyword_not ? EXPR_BEG : kw->state);
+            if (mkxp_parser_is_syntax_transform_active(p, 1, 9, -1) ? kw->id[0] == keyword_do : IS_lex_state(EXPR_BEG)) {
                 p->command_start = TRUE;
             }
             if (kw->id[0] == keyword_do) {
@@ -10614,7 +10748,7 @@ parser_yylex(struct parser_params *p)
             }
         }
         SET_LEX_STATE(IS_AFTER_OPERATOR() ? EXPR_ARG : EXPR_BEG);
-        return c;
+        return mkxp_parser_is_syntax_transform_active(p, 1, 8, -1) && c == tSTAR ? tMKXP_LEGACY_STAR : c;
 
       case '!':
         c = nextc(p);
@@ -10696,7 +10830,7 @@ parser_yylex(struct parser_params *p)
             SET_LEX_STATE(EXPR_ARG);
         }
         else {
-            if (IS_lex_state(EXPR_CLASS))
+            if (!mkxp_parser_is_syntax_transform_active(p, 1, 9, -1) && IS_lex_state(EXPR_CLASS))
                 p->command_start = TRUE;
             SET_LEX_STATE(EXPR_BEG);
         }
@@ -11038,12 +11172,18 @@ parser_yylex(struct parser_params *p)
         return '~';
 
       case '(':
+        if (mkxp_parser_is_syntax_transform_active(p, 1, 8, -1)) {
+            p->command_start = TRUE;
+        }
         if (IS_BEG()) {
             c = tLPAREN;
         }
         else if (!space_seen) {
             /* foo( ... ) => method call, no ambiguity */
         }
+        else if (mkxp_parser_is_syntax_transform_active(p, 1, 8, -1) && IS_lex_state(EXPR_ARG)) {
+            rb_warn0("don't put space before argument parentheses");
+        }
         else if (IS_ARG() || IS_lex_state_all(EXPR_END|EXPR_LABEL)) {
             c = tLPAREN_ARG;
         }
@@ -11055,7 +11195,7 @@ parser_yylex(struct parser_params *p)
         COND_PUSH(0);
         CMDARG_PUSH(0);
         SET_LEX_STATE(EXPR_BEG|EXPR_LABEL);
-        return c;
+        return mkxp_parser_is_syntax_transform_active(p, 1, 8, -1) && c == tLPAREN_ARG ? tMKXP_LEGACY_LPAREN_ARG : c;
 
       case '[':
         p->lex.paren_nest++;
